mat/vector.go

```diff
@@ -242,247 +261,362 @@ func (v *VecDense) AddScaledVec(a *VecDense, alpha float64, b *VecDense) {
+	fast := true
+	bU, _ := untranspose(b)
+	if rv, ok := bU.(RawVectorer); ok {
+		bmat = rv.RawVector()
+		if v != b {
+			v.checkOverlap(bmat)
+		}
+	} else {
+		fast = false
 	}
 
-	a, trans := untranspose(a)
-	ar, ac := a.Dims()
 	v.reuseAs(r)
 	var restore func()
-	if v == a {
-		v, restore = v.isolatedWorkspace(a.(*VecDense))
+	if v == aU {
+		v, restore = v.isolatedWorkspace(aU.(*VecDense))
 		defer restore()
 	} else if v == b {
 		v, restore = v.isolatedWorkspace(b)
 		defer restore()
 	}
 
-	switch a := a.(type) {
-	case *VecDense:
-		if v != a {
-			v.checkOverlap(a.mat)
-		}
-
-		if a.Len() == 1 {
-			// {1,1} x {1,n}
-			av := a.At(0, 0)
-			for i := 0; i < b.Len(); i++ {
-				v.mat.Data[i*v.mat.Inc] = av * b.mat.Data[i*b.mat.Inc]
-			}
+	// TODO(kortschak): Improve the non-fast paths.
+	switch aU := aU.(type) {
+	case Vector:
+		if b.Len() == 1 {
+			// {n,1} x {1,1}
+			v.ScaleVec(b.AtVec(0), aU)
 			return
 		}
-		if b.Len() == 1 {
-			// {1,n} x {1,1}
-			bv := b.At(0, 0)
-			for i := 0; i < a.Len(); i++ {
-				v.mat.Data[i*v.mat.Inc] = bv * a.mat.Data[i*a.mat.Inc]
+
+		// {1,n} x {n,1}
+		if fast {
+			if rv, ok := aU.(RawVectorer); ok {
+				amat := rv.RawVector()
+				if v != aU {
+					v.checkOverlap(amat)
+				}
+
+				if amat.Inc == 1 && bmat.Inc == 1 {
+					// Fast path for a common case.
+					v.setVec(0, f64.DotUnitary(amat.Data, bmat.Data))
+					return
+				}
+				v.setVec(0, f64.DotInc(amat.Data, bmat.Data,
+					uintptr(c), uintptr(amat.Inc), uintptr(bmat.Inc), 0, 0))
+				return
 			}
-			return
 		}
-		// {n,1} x {1,n}
 		var sum float64
 		for i := 0; i < c; i++ {
-			sum += a.At(i, 0) * b.At(i, 0)
+			sum += aU.AtVec(i) * b.AtVec(i)
 		}
-		v.SetVec(0, sum)
+		v.setVec(0, sum)
 		return
 	case RawSymmetricer:
-		amat := a.RawSymmetric()
-		blas64.Symv(1, amat, b.mat, 0, v.mat)
+		if fast {
+			amat := aU.RawSymmetric()
+			// We don't know that a is a *SymDense, so make
+			// a temporary SymDense to check overlap.
+			(&SymDense{mat: amat}).checkOverlap(v.asGeneral())
+			blas64.Symv(1, amat, bmat, 0, v.mat)
+			return
+		}
 	case RawTriangular:
 		v.CopyVec(b)
-		amat := a.RawTriangular()
+		amat := aU.RawTriangular()
+		// We don't know that a is a *TriDense, so make
+		// a temporary TriDense to check overlap.
+		(&TriDense{mat: amat}).checkOverlap(v.asGeneral())
 		ta := blas.NoTrans
 		if trans {
 			ta = blas.Trans
 		}
 		blas64.Trmv(ta, amat, v.mat)
 	case RawMatrixer:
-		amat := a.RawMatrix()
-		// We don't know that a is a *Dense, so make
-		// a temporary Dense to check overlap.
-		(&Dense{mat: amat}).checkOverlap(v.asGeneral())
-		t := blas.NoTrans
-		if trans {
-			t = blas.Trans
+		if fast {
+			amat := aU.RawMatrix()
+			// We don't know that a is a *Dense, so make
+			// a temporary Dense to check overlap.
+			(&Dense{mat: amat}).checkOverlap(v.asGeneral())
+			t := blas.NoTrans
+			if trans {
+				t = blas.Trans
+			}
+			blas64.Gemv(t, 1, amat, bmat, 0, v.mat)
+			return
```

Comment:
This needs a fallthrough, or other code for the non fast path.

Reply:
Like the one after the switch?

Reply:
Oh, sorry.